refactor: update typescript config

ilbertt opened the pull request:
# Description

Update TypeScript config to a more modern approach.

Additionally, it removes the `demos/sample-javascript` folder, which seems to be outdated.

Closes #931.

# How Has This Been Tested?

Same tests should still pass.

# Checklist:

- [x] My changes follow the guidelines in [CONTRIBUTING.md](https://github.com/dfinity/agent-js/blob/main/CONTRIBUTING.md).
- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [ ] I have edited the CHANGELOG accordingly.
- [ ] I have made corresponding changes to the documentation.

krpeacock commented:
I don't actually care that much about the jest files - it's annoying that there's any trouble with them specifically though

ilbertt reviewed (commented):
I've skipped the refactoring of `packages/bls-verify`, is it's deprecated

ilbertt commented on stats.json:
@krpeacock do we still need this file?

ilbertt commented:
Although a lot of files have changed, most of them are config files (either adapted to the new TS config or removed as not needed anymore). The source files have been changed to adapt to the new TS config: I've added `type` notation on imports, removed unnecessary comments and prefixed unused variables with `_`.

ilbertt commented:
I've tried importing the agent locally in a Node v20.18 (pre-ESM support), Node v20.19 (post-ESM support, see https://nodejs.org/en/blog/release/v20.19.0/#requireesm-is-now-enabled-by-default) and Node v22 with both `import` and `require` notations, and it seems to work properly.

nathanosdev commented on packages/agent/src/agent/http/http.test.ts:
```diff
@@ -702,7 +702,7 @@ test('should adjust the Expiry if the clock is more than 30 seconds behind', asy
   const requestBody: any = cbor.decode(mockFetch.mock.calls[0][1].body);
 
-  expect(requestBody.content.ingress_expiry).toMatchInlineSnapshot(`1260000000000`);
+  expect(requestBody.content.ingress_expiry).toMatchInlineSnapshot(`1200000000000`);
```
Why did this snapshot change?>

krpeacock commented on packages/assets/src/index.ts:
```diff
@@ -394,8 +392,6 @@ class Asset {
   constructor(
     private readonly _actor: ActorSubclass<AssetsCanisterRecord>,
     private readonly _limit: LimitFn,
-    private readonly _maxSingleFileSize: number,
-    private readonly _maxChunkSize: number,
```
Can confirm

krpeacock commented on e2e/node/basic/watermark.test.ts:
```diff
@@ -8,7 +8,7 @@ class FetchProxy {
   #replyIndex: number | null = null;
 
-  async fetch(...args): Promise<Response> {
+  async fetch(...args: Parameters<typeof global.fetch>): Promise<Response> {
```
I wasn't aware of this feature! It's convenient

krpeacock reviewed (approved):
Thanks for picking up such a hefty refactor!
